### PR TITLE
Add option for ExpansionTile to maintain the state of its children when collapsed

### LINKED ISSUE #37520: After collapsing and opening ExpansionTile, TextFormField text is deleted
## Steps to Reproduce

My ExpansionTile has several TextFormField widgets as children. When I collapse and reopen an ExpansionTile, any text entered in the form fields is deleted. I've tried adding the suggested PageStorageKey, but it still deletes entered text. Here is an example `main.dart` that reproduces the issue. Thank you!

In the sample app, 
1. Expand the tile
2. Enter any text into the field
3. Close the expansion tile
4. Open it again
5. The text is gone

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final appTitle = 'Form Validation Demo';

    return MaterialApp(
      title: appTitle,
      home: Scaffold(
        appBar: AppBar(
          title: Text(appTitle),
        ),
        body: MyCustomForm(),
      ),
    );
  }
}

// Create a Form widget.
class MyCustomForm extends StatefulWidget {
  @override
  MyCustomFormState createState() {
    return MyCustomFormState();
  }
}

// Create a corresponding State class.
// This class holds data related to the form.
class MyCustomFormState extends State<MyCustomForm> {
  // Create a global key that uniquely identifies the Form widget
  // and allows validation of the form.
  //
  // Note: This is a GlobalKey<FormState>,
  // not a GlobalKey<MyCustomFormState>.
  final _formKey = GlobalKey<FormState>();

  @override
  Widget build(BuildContext context) {
    // Build a Form widget using the _formKey created above.
    return Form(
      key: _formKey,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: <Widget>[
          ExpansionTile(
            key: PageStorageKey("test"),
            title: Text("Click to expand"),
            children: [
              TextFormField(
                key: PageStorageKey("test2"),
                validator: (value) {
                  if (value.isEmpty) {
                    return 'Please enter some text';
                  }
                  return null;
                },
              ),
            ]
          ),
          Padding(
            padding: const EdgeInsets.symmetric(vertical: 16.0),
            child: RaisedButton(
              onPressed: () {
                // Validate returns true if the form is valid, or false
                // otherwise.
                if (_formKey.currentState.validate()) {
                  // If the form is valid, display a Snackbar.
                  Scaffold.of(context)
                      .showSnackBar(SnackBar(content: Text('Processing Data')));
                }
              },
              child: Text('Submit'),
            ),
          ),
        ],
      ),
    );
  }
}
```

## Logs

```
```

```
[✓] Flutter (Channel stable, v1.7.8+hotfix.4, on Mac OS X 10.14.5 18F203, locale en-US)
    • Flutter version 1.7.8+hotfix.4 at /Users/ethan/projects/flutter/flutter
    • Framework revision 20e59316b8 (2 weeks ago), 2019-07-18 20:04:33 -0700
    • Engine revision fee001c93f
    • Dart version 2.4.0

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.0)
    • Android SDK at /Users/ethan/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS
    ✗ Xcode installation is incomplete; a full installation is necessary for iOS development.
      Download at: https://developer.apple.com/xcode/download/
      Or install Xcode via the App Store.
      Once installed, run:
        sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
    ! CocoaPods out of date (1.6.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the
        Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade:
        brew upgrade cocoapods
        pod setup

[!] iOS tools - develop for iOS devices
    ✗ Verify that all connected devices have been paired with this computer in Xcode.
      If all devices have been paired, libimobiledevice and ideviceinstaller may require updating.
      To update with Brew, run:
        brew update
        brew uninstall --ignore-dependencies libimobiledevice
        brew uninstall --ignore-dependencies usbmuxd
        brew install --HEAD usbmuxd
        brew unlink usbmuxd
        brew link usbmuxd
        brew install --HEAD libimobiledevice
        brew install ideviceinstaller
    • ios-deploy 1.9.4

[!] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[✓] VS Code (version 1.36.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.3.0

[✓] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 8.1.0 (API 27) (emulator)

! Doctor found issues in 3 categories.
```

### PR DESCRIPTION
**Pull request by eospi**
This is an updated PR for #41886 that fixes a rebase issue. Below is the original description for reference.

## Description
Currently, `ExpansionTile` rebuilds its children upon every expansion and removes the children when collapsed, resetting any state contained in the children. When building a Flutter app a couple months ago, I encountered this issue when placing a `TextFormField` inside an ExpansionTile, and any edited text would be deleted when the user collapsed the ExpansionTile. #37520 further describes the specific issue and includes sample code to reproduce it. I made this change to the `ExpansionTile` widget to make my app work correctly.
### Changes overview:
- Added a `maintainState` property to `ExpansionTile` of type `bool`
- Set `maintainState` default value to `false` for the existing behavior
- Add `!maintainState` as a condition for setting the child of the `AnimatedBuilder` to `null`

Now, an `ExpansionTile` like this maintains the state of its children while collapsed:
```dart
ExpansionTile(
  maintainState: true,
  key: PageStorageKey("test"),
  title: Text("Click to expand"),
  children: [
    TextFormField(
      key: PageStorageKey("test2"),
      validator: (value) {
        if (value.isEmpty) {
          return 'Please enter some text';
        }
        return null;
      },
    ),
  ]
)
```

## Related Issues

Fixes #37520

## Tests

I added the following tests:
- 'ExpansionTile maintainState' - creates two collapsed ExpansionTile widgets with different `maintainState` values and tests if a child widget is present in each.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by eospi** [comment]
Regarding this suggestion (https://github.com/flutter/flutter/pull/41886#discussion_r377231086), is `skipOffstage` supposed to be `false`?

File: packages/flutter/lib/src/material/expansion_tile.dart [hunk 1]
```diff
@@ -204,10 +212,11 @@ class _ExpansionTileState extends State<ExpansionTile> with SingleTickerProvider
   @override
   Widget build(BuildContext context) {
     final bool closed = !_isExpanded && _controller.isDismissed;
+    final bool shouldRemoveChildren = closed && !widget.maintainState;
     return AnimatedBuilder(
       animation: _controller.view,
       builder: _buildChildren,
-      child: closed ? null : Column(children: widget.children),
+      child: shouldRemoveChildren ? null : Column(children: widget.children),
```
Comment (Piinks) [inline packages/flutter/lib/src/material/expansion_tile.dart]: Re: https://github.com/flutter/flutter/pull/41886#discussion_r381060438
I would follow the Visibility widget and wrap with the TickerMode and Offstage widgets. I don't know that we need all of the things Visibility would add to it. :)

File: packages/flutter/test/material/expansion_tile_test.dart [hunk 2]
```diff
@@ -229,4 +229,43 @@ void main() {
...
+                  ],
+                ),
+              ],
+            ),
+          ),
+        ),
+    ));
+
+    // This text should be there while ExpansionTile collapsed
+    expect(find.text('Maintaining State'), findsOneWidget);
```
Comment (Piinks) [inline packages/flutter/test/material/expansion_tile_test.dart]: Re: skipOffstage comments from this PR and the previous one, I would add two checks here, one with `skipOffstage: true` (default) and `skipOffstage: false`, and check both cases. If it's working properly with the `TickerMode` and `Offstage` widgets, you should not find it by default, but should find it when `skipOffstage: false`.

**Review by Piinks — COMMENTED** [review]
Hey @eospi welcome! :tada: And thank you for the contribution!

**Comment by eospi** [comment]
Thanks for the suggestions! I have made the changes.

**Comment by ekremkenter** [comment]
I also need this, do you have any guesses when it will be merged?

File: packages/flutter/lib/src/material/expansion_tile.dart [hunk 3]
```diff
@@ -41,6 +41,7 @@ class ExpansionTile extends StatefulWidget {
     this.children = const <Widget>[],
     this.trailing,
     this.initiallyExpanded = false,
+    this.maintainState = false,
   }) : assert(initiallyExpanded != null),
        super(key: key);
```
Comment (Piinks) [inline packages/flutter/lib/src/material/expansion_tile.dart:46]: We'll probably want to assert that the default value has not been overridden with null.
```suggestion
       assert(maintainState != null),
       super(key: key);
```

**Review by Piinks — COMMENTED** [review]
Hey @eospi sorry for the delay, this appears to have gotten lost in my inbox. Would you like to finish this change?

**Comment by eospi** [comment]
No worries! The change is finished.

**Review by Piinks — APPROVED** [review]
Flutter_LGTM
Great! Thanks for the contribution!

**Comment by Piinks** [comment]
It looks like the customer tests are failing for reasons unrelated to this change. @dnfield, I have seen this versioning confusion appearing now on only the customer test shards... do you know how we might resolve?
I saw it on one other PR this week and ended up re-opening a new PR.

**Comment by dnfield** [comment]
Hmm. I didn't apply the tag fetching logic to analyze because it made it slow.

Maybe we could just add a tag fetch to the ci scripts for customer testing.

`git fetch https:GitHub.com/flutter/flutter --tags`

**Comment by Piinks** [comment]
WDYT, should we re-open this in a fresh pr or try resetting the git history then to resolve?

**Comment by Piinks** [comment]
@eospi I think just updating your branch with the latest from master should resolve the merge conflicts and the failing test shards.

**Comment by eospi** [comment]
It seems like a lot of unrelated changes got into the PR. I'll try to fix it in the morning!

**Comment by Piinks** [comment]
Hey @eospi, have you been able to resolve the git issue here? I've had this happen to me before. Usually I resolve it using the git reflog. Let me know if I can help resolve this. :)

**Comment by eospi** [comment]
Thanks for the suggestion! This should fix it.

**Review by Piinks — COMMENTED** [review]
@eospi It looks like those customer tests are still failing for reasons unrelated to your change. :( 
Would you be willing to open a new pull request with these changes? I would be happy to review it so we can land your contribution. Let me know, I am happy to help this get merged.

**Comment by eospi** [comment]
Thank you, that sounds good! I will open a new PR soon and let you know here when it’s done.

**Comment by eospi** [comment]
I just opened #57172 with these changes. Thank you for helping to review it!

**Pull request closed without merging**